streetmix/streetmix
# Feat: Voting API improvements

These changes enable improved functionality of the street sentiment survey. In particular,

- The `submitted` column tracks which users have submitted a vote on a particular street when using `POST api/v1/votes`. 
- Votes that have already been created can be given a `comment` by the user matching the `voter_id` value, via `PUT api/v1/votes`
- Signed-in voters will no longer receive streets they have voted on. If they have no more eligible streets to vote on, a `GET api/v1/votes` request for a vote will be given a `204` response.

## Seeding the database for testing

The new unit test adds the ability to properly test the accuracy of endpoint behavior. We create and destroy entities before and after the test runs using a specific file `app/db/seeders/votes.js`.

## QA

The following steps were taken to test this branch. To make testing easier, generate a single  vote entity manually, and ensure it has no `voter_id` value assigned.

### Testing comments + elimination of prior votes from random street selection:

* GET a vote (200)
* POST a review to this vote as a signed in user
* GET from no possible votes  (204)
* the vote record from the first step should now have the user's auth0 ID in the `submitted` field.

### Testing comments
* GET a vote (200)
* POST a review 
* PUT a comment as a signed in user
* Confirm that a user cannot PUT a comment for a vote they do not own.

### Testing removal of deleted streets from ballot pool:

* Ensure only 1 valid ballot remains to successfully GET (200)
* Find the relevant original street for said ballot and change the status to `DELETED`
* try to GET a ballot again, you should get a 204. 
* The ballot from the first step should have a `voter_id` field set to `DELETED`, this is how we denote that a street should no longer be voted on.
